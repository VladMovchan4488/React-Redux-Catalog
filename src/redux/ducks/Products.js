import { takeLatest } from "redux-saga/effects";
import { put, call } from "redux-saga/effects";

// Actions
export const GET_PRODUCTS = "GET_PRODUCTS";
export const GET_PRODUCTS_API = "GET_PRODUCTS_API";
export const GET_SEARCH = "GET_SEARCH";
export const GET_CATEGORY = "GET_CATEGORY";
export const SET_CATEGORY_VALUE = "SET_CATEGORY_VALUE";

// Default State
export const initialState = {
  input: "",
  products: [],
  category: [],
  currentCategory: "",
  pageSize: 5,
  totalCount: 20,
  currentPage: 2
};

// Dispatch
export const getProductsApi = () => ({ type: GET_PRODUCTS_API });
export const getSearch = input => ({ input, type: GET_SEARCH });
export const getCategory = category => ({ category, type: GET_CATEGORY });
export const setCategoryVAL = currentCategory => ({
  currentCategory,
  type: SET_CATEGORY_VALUE
});

// Reducer
const reducer = (state = initialState, action) => {
  switch (action.type) {
    case GET_PRODUCTS:
      return {
        ...state,
        products: action.data.products
      };
    case GET_SEARCH: {
      return { ...state, input: action.input };
    }
    case GET_CATEGORY: {
      return { ...state, category: action.category };
    }
    case SET_CATEGORY_VALUE: {
      return { ...state, currentCategory: action.currentCategory };
    }
    default:
      return state;
  }
};

// REDUX-SAGA
export function* getProducts() {
  const url = `products.json?page=${initialState.currentPage}&count=${initialState.pageSize}`;
  try {
    const data = yield call(() => {
      return fetch(url).then(res => res.json());
    });
    yield put({ type: GET_PRODUCTS, data });
  } catch (error) {
    console.error(error);
  }
}

export function* parsing() {
  yield takeLatest(GET_PRODUCTS_API, getProducts);
}

export default reducer;
