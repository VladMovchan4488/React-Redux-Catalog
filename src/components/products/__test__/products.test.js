import React from "react";
import Products from "../productList";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
import toJson from "enzyme-to-json";

configure({ adapter: new Adapter() });

it("Snaphot List", () => {
  const props = {
    products: [
      {
        img:
          "https://images-na.ssl-images-amazon.com/images/I/71Y4Ogr8BNL._SX679_.jpg",
        bsr_category: "Home & Kitchen",
        name: "A3 Magnetic Dry Wipe Weekly Planner Whiteboard for Fridge"
      }
    ],
    match: { params: {} },
    history: {
      location
    }
  };

  const wrapper = shallow(<Products.WrappedComponent {...props} />);

  expect(toJson(wrapper)).toMatchSnapshot();
});
