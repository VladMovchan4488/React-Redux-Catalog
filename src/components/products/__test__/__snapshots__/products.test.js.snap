// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snaphot List 1`] = `
<ForwardRef
  noGutters={false}
>
  <Col
    key="0"
    lg={4}
    md={6}
    xs={12}
  >
    <Product
      img="https://images-na.ssl-images-amazon.com/images/I/71Y4Ogr8BNL._SX679_.jpg"
      name="A3 Magnetic Dry Wipe Weekly Planner Whiteboard for Fridge"
    />
  </Col>
</ForwardRef>
`;
